ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,899 @@
+{
+    for (int sig = 1; sig < NSIG; ++sig) {
+        // these signals can't be overridden
+        if (sig == SIGKILL || sig == SIGSTOP)
+            continue;
+
+        // Set signal dispositions.
+        // This avoids inherit unwanted overrides.
+        // Also prevent handling unwanted signal handler, especially using vfork().
+        // Use ignore SIGPIPE for compatibility with forkexecd.
+        signal(sig, sig == SIGPIPE ? SIG_IGN : SIG_DFL);
+    }
+}
+
+static void
+clear_cgroup(void)
+{
+    int fd = open("/sys/fs/cgroup/systemd/cgroup.procs", O_WRONLY|O_CLOEXEC);
+    if (fd >= 0) {
+        char string_pid[32];
+        int ignored __attribute__((unused));
+        sprintf(string_pid, "%d\n", (int) getpid());
+        ignored = write(fd, string_pid, strlen(string_pid));
+        close(fd);
+    }
+}
+
+CAMLprim value
+caml_pidwaiter_dontwait(value waiter_val)
+{
+    CAMLparam1(waiter_val);
+    PIDWAITER(waiter, waiter_val);
+
+    pthread_mutex_lock(&waiter->mtx);
+    bool reaped = waiter->reaped;
+    waiter->reaped = true;
+    pthread_mutex_unlock(&waiter->mtx);
+
+    // reap the pid to avoid zombies
+    if (!reaped) {
+        pthread_t th;
+        if (create_thread_minstack(&th, thread_proc_reap, (void *) (intptr_t) waiter->pid) == 0)
+            pthread_detach(th);
+    }
+
+    CAMLreturn(Val_unit);
+}
+
+typedef struct {
+    pid_t pid;
+    bool timed_out;
+    bool stop;
+    struct timespec deadline;
+    pthread_mutex_t mtx;
+    pthread_cond_t cond;
+} timeout_kill;
+
+static void *
+thread_proc_timeout_kill(void *arg)
+{
+    timeout_kill *tm = (timeout_kill *) arg;
+
+    pthread_mutex_lock(&tm->mtx);
+    int res  = tm->stop ? 0:
+        pthread_cond_timedwait(&tm->cond, &tm->mtx, &tm->deadline);
+    pthread_mutex_unlock(&tm->mtx);
+
+    if (res == ETIMEDOUT) {
+        kill(tm->pid, SIGKILL);
+        tm->timed_out = true;
+    }
+    return NULL;
+}
+
+static int
+create_thread_minstack(pthread_t *th, void *(*proc)(void *), void *arg)
+{
+    int res;
+
+    // disable any possible signal handler so we can safely use a small stack
+    // for the thread
+    sigset_t sigset, old_sigset;
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    pthread_attr_t th_attr;
+    res = pthread_attr_init(&th_attr);
+    if (res)
+        return res;
+    pthread_attr_setstacksize(&th_attr, PTHREAD_STACK_MIN);
+
+    res = pthread_create(th, &th_attr, proc, arg);
+
+    pthread_attr_destroy(&th_attr);
+    pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
+
+    return res;
+}
+
+/*
+ * Wait a process with a given timeout.
+ * At the end of timeout (if trigger) kill the process.
+ * To avoid race we need to wait a specific process, but this is blocking
+ * and we use a timeout to implement the wait. Timer functions are per
+ * process, not per thread.
+ * Returns <0 if error, 0 if not timed out, >0 if timedout.
+ */
+static int
+wait_process_timeout(pid_t pid, double timeout)
```

Comment:
The main reason is that the function to convert wait result from C to Ocaml is accessible from C. I tried to split the reap and result conversion (calling wait from Ocaml) but to avoid racing updating reap flag to avoid reaping potentially twice was complicated and wrong. It's much easier to avoid split implementation between Ocaml and C and handle everything in a single place.